source/1_engine/rex_engine/src/cmdline.cpp

```diff
@@ -0,0 +1,205 @@
+            continue;
+
+          if(rsl::strincmp(arg.data(), active_arg.argument.data(), arg.length()) == 0)
+          {
+            return active_arg.value;
+          }
+        }
+
+        return rsl::nullopt;
+      }
+
+    private:
+      void parse_cmd_line(rsl::string_view cmdLine)
+      {
+        // all we have to do is split the command line based with spaces
+        // there's no need to perform any heap allocation for that.
+        count_t start_pos = cmdLine.find_first_not_of(' ', 0); // it's possible, although unlikely, that we start with a space
+
+        // all executables should be called with at least 1 argument
+        // the first argument should always be the path to the executable itself
+        REX_ASSERT_X(start_pos != -1, "No arguments given to your exe, there should always be at least 1 arg.");
+
+        // the module we're launching is always the first argument on the commandline
+        // we skip this
+        count_t space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        // if we have no arguments, we just return
+        // this is possible when running without a debugger
+        if (space_pos == -1)
+        {
+          return;
+        }
+
+        start_pos         = space_pos + 1;
+
+        // get the new space pos
+        space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        const rsl::string_view arg_prefix = "-"; // all arguments should start with a '-'
+        while(start_pos != -1 && space_pos != -1)
+        {
+          const count_t length            = space_pos - start_pos - arg_prefix.size();
+          const rsl::string_view argument = cmdLine.substr(start_pos + arg_prefix.size(), length);
+          add_argument(argument);
+          start_pos = cmdLine.find_first_not_of(' ', space_pos); // skip all additional spaces
+          space_pos = cmdLine.find_first_of(' ', start_pos);
+        }
+
+        if(start_pos != -1)
+        {
+          add_argument(cmdLine.substr(start_pos + arg_prefix.size())); // + 1 to ignore '-'
+        }
+      }
+
+      void add_argument(rsl::string_view arg)
+      {
+        const count_t equal_pos = arg.find('=');
+        rsl::string_view key    = "";
+        rsl::string_view value  = "";
+
+        // if the argument is of type -MyArg=Something
+        if(equal_pos != -1)
+        {
+          key   = arg.substr(0, equal_pos);
+          value = arg.substr(equal_pos + 1);
+        }
+        // if the argument is of type -EnableSomething
+        else
+        {
+          key   = arg;
+          value = "1"; // this is so we can easily convert to bool/int
+        }
+
+        auto cmd_it = rsl::find_if(g_command_line_args.cbegin(), g_command_line_args.cend(), [key](const Argument& cmdArg) { return key == cmdArg.name; });
+
+        if(cmd_it == g_command_line_args.cend())
+        {
+          REX_WARN(LogEngine, "Command {} passed in but it's not recognised as a valid command so will be ignored", arg);
```

Comment:
Should we maybe return the invalid one such as in Python? Not sure if it's useful though.

Reply:
not sure what we gain with that, we just ignore it and print a warning about it